core/types.ml

```diff
@@ -143,6 +176,7 @@ type typ =
     | `Lens of lens_sort
     | `Alias of ((string * type_arg list) * typ)
     | `Application of (Abstype.t * type_arg list)
+    | `RecursiveApplication of rec_appl
```

Comment:
I was hoping to use `Application` rather than introducing `RecursiveApplication`. Perhaps it matters not here, but for the module system, I think it shouldn't be necessary as we'll resolve names. I think we will try to unify the two in a subsequent module patch.

Reply:
As it stands, `RecursiveApplication` has a lot more structure (for example, dualisation, unwinding, linearity tracking amongst a group, etc.) than `Application` at the moment. Maybe a better design would unify `Application` and `RecursiveApplication` to something like:

```
type application = | Abstract of Abstype.t | RecursiveApplication of rec_appl
...
| `Application of application * type_arg list
```

Reply:
I did not mean to imply that `Application` would suffice as is. In time we'll get rid of the special handling of abstract types too. The next iteration of the module system is likely to introduce abstract types.